### PR TITLE
Ed25519 signatures, initial implementation

### PR DESCRIPTION
**Pull request by olorin**
On top of #33 ([diff](https://github.com/ambiata/tinfoil/compare/topic/less..topic/gadt)).

Context: bindings to `libsodium` (as a shared library) implementing keypair generation, message signing and verification with `ed25519`. This is for use in the prototype implementation of the asymmetric signing protocol being developed in `zodiac`. `ed25519` was chosen for reasons of performance/key size and relative resistance to side-channel attacks.

Tried to do something a bit more flexible with the key/signature types (GADT parameterised on algorithm), but I'm not totally sure it's the best way to do this. Also the first time I've had a reason to use GADTs, so feedback would be appreciated.

The version of `libsodium` packaged for Centos 6 is a bit old and doesn't support detached signatures, so this is emulated in `tinfoil` (works with newer versions of the library too of course).

This PR also removes protocol drafts which would be better off in zodiac.

@markhibberd 

/cc @erikd-ambiata @thumphries

### CODE REVIEW

**Comment by erikd-ambiata** [comment]
Its a lot of code, but I didn't see anything amiss.

File: src/Tinfoil/Signing/Ed25519/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+import           System.IO (IO)
+import           System.IO.Unsafe (unsafePerformIO)
+
+import           Tinfoil.Data.KDF
+import           Tinfoil.Data.Key
+
+-- | Generate a new Ed25519 keypair.
+--
+-- FIXME: init requirement for thread safety?
```
Comment (thumphries) [inline src/Tinfoil/Signing/Ed25519/Internal.hs]: Could you elaborate on this?

Comment (olorin) [inline src/Tinfoil/Signing/Ed25519/Internal.hs]: Indeed, I forgot I hadn't looked at that yet, thanks. Hold on, will work out whether this is an issue or not.

Comment (olorin) [inline src/Tinfoil/Signing/Ed25519/Internal.hs]: Right, yep, initialization is required for older versions of `libsodium` than the one I'm developing with locally, which I think is why the tests are failing on `ci` now. Will fix this.

**Comment by thumphries** [comment]
> Tried to do something a bit more flexible with the key/signature types (GADT parameterised on algorithm), but I'm not totally sure it's the best way to do this.

I like the phantom type here, seems appropriate. Doesn't seem to be buying us anything right now, though, since all the use site are monomorphised to `Ed25519`. Will you be adding any typeclasses or polymorphic functions over `PublicKey` or `SecretKey` eventually? Would be a nice API if so.

Sodium bindings look good, based on the QuickCheck coverage. This being `tinfoil` you should of course take my opinions with a ladle of NaCl.

🍧

**Comment by olorin** [comment]
> Will you be adding any typeclasses or polymorphic functions over PublicKey or SecretKey eventually? Would be a nice API if so.

Yeah, I figured polymorphic `verify` and signature-parsing and things can be exposed from `Tinfoil.Signing` if other algorithms are implemented. Only reason I'm unsure about the interface is I'm not sure how likely we are to ever implement an alternative to ed25519.

File: src/Tinfoil/Signing/Ed25519/Internal.hs (hunk 1 above)
Comment (olorin) [inline src/Tinfoil/Signing/Ed25519/Internal.hs]: Nope, tests were failing due to another version difference. Initialization has to do with initializing the RNG and other such things - also important, so this should be fixed before this is merged even though the tests are passing now.

Comment (olorin) [inline src/Tinfoil/Signing/Ed25519/Internal.hs]: @thumphries 

> Initialization has to do with initializing the RNG and other such things - also important, so this should be fixed before this is merged even though the tests are passing now.

On closer examination, this doesn't do as much as I thought - the default PRNG doesn't need explicit seeding, and is also thread-safe before and after `sodium_init`, along with the default implementation of all the other functionality in `libsodium`; the only operation which is not thread-safe is `sodium_init` itself. So I think this PR is good as-is and I'll expose the init functionality later on if there's benefit to doing so.

Comment (thumphries) [inline src/Tinfoil/Signing/Ed25519/Internal.hs]: 🍍  thanks for following up

**Comment by olorin** [comment]
Superseded by #42, must have missed a rebase somewhere.

**Pull request closed without merging**